### DISCUSSION TITLE
Skipping Nullable parameters in constructor for GuardClauseAssertion.

### DISCUSSION THREAD

**Discussion by PatrickGhosn** [AutoFixture/AutoFixture#1377] [Q&A] [answered]
Hey all,

We're extending the default IBehaviorExpectation, and so far we've successfully had it skip optional parameters and certain Nullable types such as string, etc.

However we are not able to make it work when we're using it to test our own classes when they are Nullable. We've tried several ways, but none seem to work.

For reference, this is what we're using to have it check the working Nullables:

```         
            .........
            if (checkCommand != null && IsNullable(checkCommand.ParameterInfo.ParameterType)) { return; }

            origin.Verify(command);
        }

        private static bool IsNullable(Type type)
        {
            return (Nullable.GetUnderlyingType(type) != null);
        }
````

For things such as `string?` it's working perfectly fine. But when it tests our own classes, it finds that it's not Nullable. We also tried the below, however that made it skip everything including the non nullable fields.
```
        private static bool IsNullable(Type type)
        {
            if (!type.IsValueType) return true;
            return (Nullable.GetUnderlyingType(type) != null);
        }
```

Any help or hint on what we're doing wrong would be appreciated!

**Comment by GeraldLx** [comment, marked as answer]
I guess you are not really understanding what string? means. There is a huge difference between string? and int?.

They address two totally different things and look from a syntactic perspective the same.

When you pass in a value type, like an integer, and you get the default value, you cannot know if this is intentional or by accident.

If 0 is a valid value you cannot write code like this:
```
public void Foo(int number)
{
  if (number == 0)
  {
    throw ...
  }
}
```
So with .NET 2 came the new class Nullable<T> to address this.
Now you could differ between null and 0.

At the same time C# 2 came out and provided a convenience alias of int? for Nullable<int>.

So the goal was to allow null for value types.

Now with C# 8 we got the nullable feature for reference types. 
The feature is made to allow the compiler to know if null is an allowed value.

public void Foo(string text) Null is not allowed
public void Foo(string? text) Null is allowed

While the first feature is in .NET (the Nullable class), the second is a pure C# compiler feature (the questionmark here is just an info marker for the compiler, there is nothing in the runtime to support this).

Under the hood the C# compiler just adds an attribute to the parameter. So your nullable check is wrong. You dont want to test, if the parameter is a nullable, you want to test if the parameter is tagged with the attribute.

You can solve this by going low level or by using the NullabilityInfoContext.

Here a small demo code, to summarize what I wrote:
```
internal static class Program
{
  public static void Main(string[] args)
  {
    var type = typeof(Sample);
    var method = type.GetMethod(nameof(Sample.Foo));

    foreach (var parameter in method!.GetParameters())
    {
      Console.WriteLine(IsNullable(parameter));
    }
  }

  private static bool IsNullable(ParameterInfo parameter)
  {
    var context = new NullabilityInfoContext();
    var info = context.Create(parameter);

    return info.WriteState is NullabilityState.Nullable;
  }
}

public class Sample
{
  public void Foo(string text1, string? text2, int number1, int? number2)
  {
  }
}
```
Back to AutoFixture and the GuardClauseAssertion. By the time this was written, C#8 was not out. So of course the code cannot detect this new feature.
I guess future version will have support for this, until now you can only change the behavior, like you do. Just fix your IsNullable check and it should work

**Reply by PatrickGhosn** [reply]
Hey! Yes after my question I realized that string is different from int and our behavior was not working as intended.

As for the solution you proposed, sadly it looks like the solution we have uses .NET 5.0, so we don't have access to NullabilityInfoContext. But it shows at least that we can't solve this issue without updating the version  so we'll do that for the tests project.

Thank you for the great answer. :)
